### ISSUE TITLE
Preview fails to display edits in Gutenberg

### ISSUE THREAD

**Issue by Armarsh** [WordPress/gutenberg#13232]
**Describe the bug**
The preview button in the Gutenberg editor in Wordpress 5 RC3 doesn't display unsaved edits.

Reproduced by creating and saving a page, then edit the page and click preview. The preview page only shows the original page and not the additional edits.

(Same as issue #12617 which was closed -- but this is NOT a plugin issue and should NOT be closed or ignored based on the fact that the problem also can be reproduced with some plugin conflicts.)

**Comment by designsimply** [comment]
Noting that #12617 had some speculation come up in the comments and a plugin conflict was tested and confirmed and that opening a new issue with the relevant details for your case is a good route. I'm copying over some relevant information from your comment at https://github.com/WordPress/gutenberg/issues/12617#issuecomment-448446822 for reference:

> You could be on to something there with the comment about transient or DB delay. I have about 10 sites on a shared VPS server -- and I have found only 2 sites that don't have this problem. Those 2 are very low traffic sites as compared to the others -- so it could definitely be something that is tied to database contents & load time. That would also explain why it might not show up in new, clean testing installations.
> 
> Edit to add: I took a look at the size of the database for each of my sites. The two sites that function well with previews have database sizes of 3M or under. Among the sites that are giving me problems, the smallest DB is 4.1M -- but most are larger. I wouldn't expect my particular values to hold true for other sites, as it would be a server-side issue that would also be tied to server load and resources. But at least this gives weight to the DB delay theory.

**Comment by designsimply** [comment]
The conversation at https://wordpress.slack.com/messages/C02QB8GMM/ also mentions databases and heavy caching:

> I would not be surprised to have it happen more on .org because of the fact that we use multiple databases and heavy caching, so there can be a delay between saving content and having it appear back on a reading of that content, but that delay tends towards the very small.
> I think the preview window tries to load a little bit too quickly. This may be a problem new to 5.0

**Comment by talldan** [comment]
> Reproduced by creating and saving a page, then edit the page and click preview. The preview page only shows the original page and not the additional edits.

@Armarsh I couldn't reproduce this, if you could provide some details reproduction steps that'd be really helpful. Here's the capture of my test:
![testing-preview](https://user-images.githubusercontent.com/677833/51012976-f8831d80-159a-11e9-9aac-495d2e7022a0.gif)

**Comment by Armarsh** [comment]
This problem was already described at length in #12617  by multiple posters -- the problem was pretty clearly described by many.   I opened this thread after #12617 was retitled and closed because of a separate plugin conflict that was not tied to or mentioned in the original report. 

There is also a demo posted  #12717  - see this link https://www.youtube.com/watch?v=FhCssW8JpEc&feature=youtu.be

@talldan -- here's my question for you:  what is the size of the database for the wordpress installation you are using for testing?  If you are using a clean, new installation then you aren't going to see the problem.  So I think it's best tested on a well-established, larger site.

**Comment by Chrico** [comment]
Howdy,

i can reproduce this behavior as well, but just for `post_status = 'private'`. As soon as i switch to `public` and add elements, the preview is working. If i switch back to `private` and add elements, they will not appear in preview mode.

**Comment by Chrico** [comment]
Sry for commenting twice..on my local machine the error is not there, but on stage and production it's there (same Server, PHP, Database, Plugins, ...). 
It seems, that this error is not present when dealing with "slower" machines. I'll try to remote debug this now, but this may can take a while until i can get a enviornment where i can rebuild it. 

Locally in `wp-includes/revision.php` in `_set_preview`-function uses already the correct post, so it's stored in DB before the preview is rendering.

What i can see so far is, that in `PostPreviewButton`-Component the `openPreviewWindow` [1] maybe should wait until the "savePost" or `autosave` is finished before loading the preview. Since this is running async [1] and _after_ opening/refreshing the preview window [2] it may can happen that the preview is rendered, before the actual post is available for previewing.



[1] https://github.com/WordPress/gutenberg/blob/2575d239c586a45c86935dd0b07e11e2b1ca00ca/packages/editor/src/components/post-preview-button/index.js#L132

[2] https://github.com/WordPress/gutenberg/blob/2575d239c586a45c86935dd0b07e11e2b1ca00ca/packages/editor/src/components/post-preview-button/index.js#L157-L163

**Comment by designsimply** [comment]
Based on the information in https://github.com/WordPress/gutenberg/issues/12617#issuecomment-449520938 I ran the following test:

1. Go to https://jurassic.ninja/ and create a new test site.
1. Go to Plugins > Add New, search for FakerPress, install & activate.
1. Go to FakerPress > Posts and generate 200 posts and 80 pages.
1. Check the size of the database using `wp db size --size_format=kb` via `wp-cli` (ssh credentials for jurassic.ninja sites are provided in an admin notice).
1. Go to Posts and open any published post.
1. Make any edit to the post.
1. Click the "Preview" button and check to see if the edit appears in the preview.

**Result:** the database including the sample content was 5.8 MB and all previews still worked normally in my own testing. ([1m38s](https://cloudup.com/cOSeeEj2dL5))

**Comment by designsimply** [comment]
Based on the information in https://github.com/WordPress/gutenberg/issues/13232#issuecomment-453511767 I ran the following test:

1. Open a previously published post.
1. Go to Settings > Document > Status & Visibility and change "Public" to "Private".
1. Make any edit to the post.
1. Click the "Preview" button and check to see if the edit appears in the preview.

**Result:** previews to previously published private posts work normally for me in testing.

**Comment by Armarsh** [comment]
@designsimply - thanks for trying to test this -- but I do think that if it is tied in any way to database response time then it is going to vary depend on server configuration and resources.   So maybe a site running in a shared hosting environment with one provider is going to have the problem, and if the same site were moved to a different server it won't have the problem anymore.  And in a shared hosting environment, there are also going to be problems completely outside the control of the tester, such as the impact of resource usage from other users on the system.   

Maybe you can keep on trying on jurassic.ninja to push until you get to whatever is enough to trigger the problem -- or maybe that site has a configuration that is better optimized for wordpress than the hosts and servers used by those of us reporting the problem. 

So probably the best way to test is on a server that is known to have this issue.

**Comment by talldan** [comment]
Hi @Chrico Thanks for looking into that. The post preview button should handle the asynchronous nature of the save.

The steps that happen are:
1. The preview window/tab is opened if required
https://github.com/WordPress/gutenberg/blob/ad4c9e745d1b4f3bca1f38bf4e8274967b345542/packages/editor/src/components/post-preview-button/index.js#L140-L143

2. If the post is autosaveable, it is saved (or if not autosaveable, the current post is shown and the following steps don't happen)
https://github.com/WordPress/gutenberg/blob/ad4c9e745d1b4f3bca1f38bf4e8274967b345542/packages/editor/src/components/post-preview-button/index.js#L150-L163

3. A loading screen is shown in the preview window
https://github.com/WordPress/gutenberg/blob/ad4c9e745d1b4f3bca1f38bf4e8274967b345542/packages/editor/src/components/post-preview-button/index.js#L165-L167

4. When REST request to save the post is complete, the PostPreviewButton component `previewLink` prop is updated and the `PostPreviewButton` `componentDidUpdate` method updates the preview window to show the latest post
https://github.com/WordPress/gutenberg/blob/ad4c9e745d1b4f3bca1f38bf4e8274967b345542/packages/editor/src/components/post-preview-button/index.js#L102-L125

Are either you or @Armarsh using metaboxes when you experience the issue? They very slightly change how the preview button works.

**Comment by Armarsh** [comment]
@talldan -- I'm not sure what a metabox would be or include.  Can you give me an example on the simplest level? (that is, something that would commonly be used in a simple theme).

**Comment by talldan** [comment]
They're a way to add extra metadata to a post. A plugin might add extra metadata/metaboxes. They can appear as extra options in the sidebar or at the bottom of the editor.

Custom Fields are the most accessible way to add metadata (enabled in Gutenberg in the Options dialog, accessible from the menu in the top right corner).

My understanding is that Gutenberg has no way of identifying when a metadata value has changed and needs to be saved, so the behaviour of the preview button changes to save every time it's clicked, just in case there is an unsaved metadata change.

I've just done a run-through, and using a custom field is the only way I've reproduce this issue yet. Here's the steps:
1. Start a new post
2. Open the options dialog from the Editor 'more' menu. (The menu accessible from the top right corner, icon looks like the three stacked dots)
3. Enable the Custom Fields option
4. Add a new custom field
5. Add a title and content to the post and save
6. Add some extra text to the post content and click the preview button
7. Observe that the changes made are not reflected in the preview

I wonder if that's main culprit.

**Comment by Armarsh** [comment]
I have 9 sites, all on the same shared server, showing the problem.  

I have been doing checking while following this thread on one of the more simple sites.   This site does not have any custom fields activated. 

I do have some other sites that have custom fields  --- but I have been doing the testing for this function on the site that has the simpler structure. 

I'd note that I don't believe I have the Gutenburg custom fields in use on any of these sites -- these are all well-established sites so not really creating anything new.  However, some of the sites do have custom fields created with different plugins.  But again, I've tested on all sites on my server - and the process to create the error is the same for all.

**Comment by talldan** [comment]
Are you able to let us know what plugins you have active on the site with the simpler structure? I think both @designsimply and I have been able to establish that the bug doesn't happen with a fresh install, it'd be great to narrow this down.

**Comment by Armarsh** [comment]
But the problem persists with all plugins deactivated --- I just tested once again to confirm that.  

So I just don't think it's a plugin issue. 

I'd note that if install the  Health Check & Troubleshooting Plugin and run the troubleshooter, the problem goes away --- even with all plugins enabled --- but that plugin is creating some sort of shadow session that is running separately.

**Comment by Chrico** [comment]
I think the solution is posted here: https://github.com/WordPress/gutenberg/issues/12617#issuecomment-453917507 - seems that, when having Metaboxes, the save routine does not work correctly when clicking "preview".

**Comment by v33dub** [comment]
Just wanted to comment that I have been experiencing this issue - I have several sites on the same server, the much smaller 1-5 page sites have no problem previewing, but the larger sites do. None of my sites are using Metaboxes, so I don't think it's related.

**Comment by kristarella** [comment]
I was able to replicate this issue on a number of sites and for one of them I tried clearing the cache (via WP CLI) and reloaded the preview, which revealed the change in the preview (while public page is still showing the previous published version.

This could be one reason that the problem is hard to reproduce: different sites and servers have different caching settings.

I'm not sure what further information is helpful for this, so if you would like more info for setup of the sites I was testing please ping me.

**Comment by retnonindya** [comment]
Hello! A user contacted us, reporting about this issue and I was able to replicate it both on the user's site and on my test site. In case you want to know more on the info for the user's website and my test sites, feel free to ping me.

**Comment by davoraltman** [comment]
I was able to reproduce the issue as well.

zd-1759856

**Comment by khristiansnyder** [comment]
I was able to reproduce as well. hc-9015982.

**Comment by rantoncuadrado** [comment]
zd-1759856 is waiting for updates :)

**Comment by locksoft** [comment]
Could it be related to the browser caching system?
I always have this issue with Firefox, that is my default browser, but everything seems working well on Safari.

**Comment by Armarsh** [comment]
Bad news -- upgrade to WP 5.1 makes the problem WORSE.   The preview still fails to display under the conditions outline in the original post -- but the workaround of refreshing or reloading the preview page no longer works.  I've tested this on 2 sites.  So now no preview function at all on previously published pages.

**Comment by earnjam** [comment]
@Armarsh I just want to confirm, you've been able to replicate the issue consistently with no plugins active and set to the default theme?

**Comment by marcber** [comment]
I made this comment in #12617, maybe it can be use for those reading this item...

--------------------------------------
    * it only happens on specific existing pages (created new ones and the bug was not present)

    * for the pages where the bug shows up, the browser only loads the page once

    * for the pages where the bug is NOT present, the browser loads the page content twice. (This is actually not very clean of a design, even in the cases where it "works". You can see the current published content show up a fraction of a second... then the page goes blank, the browser initiate a second load and the current non saved draft edits show up finally).

So can we assume that it is related to some states related to a given page post within the SQL DB, which are used in the preview button logic? Those states being also linked to the auto-save logic seems a safe bet indeed.

The whole preview logic should maybe be rewritten from the ground up... I've tracked issues related to this back from 2013 or so, on pretty old WordPress releases. It's been a rampant bug for quite a while it seems.
----------------------------------

I have no idea of I'm using metaboxes or not. I'm coming from an Elementor (free version) setup. What seems clear is the double load of pages symptom. Worse, I've seen occurrences where the page initially double load... then a few minutes later, will reload once again. It's like if the code is so unclear about what should be shown, that it just basically keep on reloading, based on a timer, what seems to be the latest sources of a page. It's a brute force method to me, trying to avoid a more logical approach, where we would know right away, what exactly should be displayed, once. If later on, the user makes further edits to a draft / pending post (that has a published version online), he should simply refresh the preview page manually in his browser... There should not be some java script running with a timer or some sort of logic in the background, inside the generated preview page.

My feeling is that, by trying to fix, and re-fix over it, as time goes, it became pretty messy in the way it handles the previewing in general. By first assessing the full picture of what the preview code does, THEN trying to re centralize the whole thing in a single place in the code, where all limit cases are handled cleanly, then we should get something which is stable and secure. The current case is probably that the logic related to previewing is all over the place in the code various modules, thus making it virtually impossible to debug and maintain for all cases.

**Comment by earnjam** [comment]
> There should not be some java script running with a timer or some sort of logic in the background, inside the generated preview page.

That's not how it works. It gets refreshed anytime a save occurs in the editor. There isn't any logic happening in the preview window. 

The logic to open/refresh the preview from Gutenberg is all contained inside the [`PostPreviewButton`](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/post-preview-button/index.js) component.

**Comment by marcber** [comment]
Humm, then, how can we explain that, while I'm looking at the preview page, it reloads by itself, without the user doing anything at all, a good 1 minute after that page was first generated ? I've seen this on Firefox.

**Comment by earnjam** [comment]
From an Autosave firing in the editor. Close your editor window and the preview window will no longer refresh.

**Comment by marcber** [comment]
This autosave thing also creates problems related to the current issue from other threads I've seen somewhere.... It's all linked. If an autosave fires, it shouldn't necessarily trigger a reload of a preview page... not unless there is actually a change. Is there an internal flag that gets change to "new" when the user edits a page content ? The same logic goes that an autosave shouldn't add data in the database, unless there's actual changes. So, we could maybe say that saving new data into the DB and sending to a preview page that is open are in sync logic wise? But currently, it doesn't look like that. The preview seems to reload, even when it's not required.

**Comment by Armarsh** [comment]
> @Armarsh I just want to confirm, you've been able to replicate the issue consistently with no plugins active and set to the default theme?

I did that before my initial report -- disabled all plugins & set default theme to test.   I did not repeat that after reporting the changes post 5.1, in part because others reported similar issues.  As noted, this is consistent across most of my sites on a shared VPS server, with a variety of different themes & plugins.

**Comment by earnjam** [comment]
Ok, well the reason I ask is this comment.

>But the problem persists with all plugins deactivated --- I just tested once again to confirm that.
>
>So I just don't think it's a plugin issue.
>
> I'd note that if install the Health Check & Troubleshooting Plugin and run the troubleshooter, the problem goes away --- even with all plugins enabled --- but that plugin is creating some sort of shadow session that is running separately.

The [Health Check & Troubleshooting plugin's troubleshooting mode](https://make.wordpress.org/support/handbook/appendix/troubleshooting-using-the-health-check/#troubleshooting) deactivates all plugins and uses a default theme for requests from the current user. 

Some themes add metaboxes to the editor as well. It's possible if you only deactivated plugins in your testing (or didn't do both plugin and the theme at the same time), then your theme may be adding metaboxes to the editor, which would be triggering the problem identified in #12617.

**Comment by marcber** [comment]
> 
> The logic to open/refresh the preview from Gutenberg is all contained inside the [`PostPreviewButton`](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/post-preview-button/index.js) component.

I've checked briefly this code, with my limited knowledge of Java. It seems to me that this script does very little, if not nothing, in terms of actual decision logic. It merely creates a preview tab/page, then maintain the links to it. It seems, as I expected, that various external modules send "render" requests to this script, in any given order they happen to come. This is where the problem lies. If I'm right with this, it means that this script _does not_ centralize at all the actual management of the previewing mechanics. 

A question:

Do we actually need the preview system to be that sophisticated? By that I mean, do we need preview to be refresh-able like it is right now? Honestly, the first time I clicked on the preview button, I only expected it to open to a new windows/tab and render a snapshot of my current latest version of a document I'm working on.

If we would simplify all this "logic" (that is spread across many plugins, from various coding teams, etc) and just do such that a preview render ONCE, based on the latest saved (or auto-saved) version of a document, if the user later-on makes edits, HE would be responsible to hit F5 or whatever is used to refresh the preview page on his particular browser.

This whole auto-refresh thing is making things much more complex than it really needs to be, for what seems to be a simple feature at the base.

Moreover, the preview object (it should be more OOP oriented) should be autonomous in making decisions on what to do regarding what to render inside the preview page... no matter if it receives 5 or more different "render" requests, it is the responsibility of that preview object to make the final call, depending on its knowledge of the state of the content freshness, etc.

**Comment by kristarella** [comment]
I heard that the Preview code has changed, and in my latest tests I can no longer replicate this problem where I could before. Can anyone else confirm the symptoms are resolved for them?

**Comment by talldan** [comment]
@kristarella Yep, in my testing it seems to be resolved 🎉 

Would love to know what the change was that resulted in the fix.

**Comment by Armarsh** [comment]
NO FIX FOR ME -- I'm seeing the same problem with WP 5.1.1. as reported above. Preview function does not work for published posts.

**Comment by talldan** [comment]
@Armarsh There's a good chance the fix hasn't made it into a release yet. The next release for the plugin should go out on Wednesday.

**Comment by kristarella** [comment]
@Armarsh Thanks, my test was not thorough enough earlier. Just tried again with three different sites, with new and old published pages and edits do not show.

Any **draft** shows preview normally. Any **published** or **private** post does not show edits in the preview, but switching from **published to private**, or **private to published**, or switching to draft causes the edits to show in preview.

And as you noted earlier, clearing cache and reloading preview no longer works as it did for me in previous versions.

**Comment by kristarella** [comment]
@talldan Not sure it is fixed actually. Which version are you testing?

**Comment by Armarsh** [comment]
@talldan What plugin are you referring to?  This is a core WP problem for me -- multiple sites, and tested before I opened this thread by disabling all plugins & reverting to TwentyNineteen theme.

**Comment by earnjam** [comment]
@armarsh this plugin. Development of the editor occurs here and then is merged into core at each release. I have not tested yet, but @talldan is saying the bug is no longer occurring in the latest on the master branch of this repo. It’s not in core yet.

**Comment by Armarsh** [comment]
@earnjam - Ok, thanks for the explanation.  Is there a link where we non-developers can download the pre-release plugin to test?

**Comment by earnjam** [comment]
From the main page on this repo, there is a link to download a zip file, but you will need npm installed to build it in order for it to work. 

Once the next plugin version is released (which isn’t too far off), you will be able to install/test it from the Wordpress.org plugin repository like any other plugin.

**Comment by talldan** [comment]
@kristarella testing in a development environment using the code checked out from the repo. It would be great to get some further confirmation, hopefully when the plugin is updated next week or in the upcoming 5.2 release.

**Comment by talldan** [comment]
After testing again, it looks like the problem still exists 😞 

Sorry to get everyone's hopes up.

**Comment by marcber** [comment]
It's even maybe getting worse. Just checked on our live site and now, even for the end-users, they see a (forced) double refresh of every pages after they load.

This whole scheme of forcing a reload (just in case / to be sure) thing needs to go for good. It's not a clean way to handle this.

We need an internal function that can be called to render a page... _once_, with the right content on the first shot at it, no matter the circumstances. That function needs to be self-aware and decide what content is right to render... No matter how many developers, in how many plugins etc makes bazillions of simultaneous requests to render a given version of a page.

Does that makes any sense to anyone here or I'm just totally off track ??

Regards

**Comment by earnjam** [comment]
I've mentioned before, but that's not how it works.

**Comment by locksoft** [comment]
Still no solutions?

**Comment by locksoft** [comment]
Still no solutions?

**Comment by youknowriad** [comment]
Based on this comment https://github.com/WordPress/gutenberg/pull/14877#issuecomment-481881683 and #14877 I think this can be closed now, right?

**Comment by talldan** [comment]
Yep, I'll close it, we can always reopen it if it does continue to be an issue. I'm hopefully looking forward to this one being fixed for everyone, though 👍

**Comment by arnisjuraga** [comment]
I had a problem with WordPress 5.1.1 and ACF (v.5.7.12) plugin. With ACF enabled and one custom field defined for "post" type, preview is working till first live Publish.
Editing posts afterwards and pressing "Preview" button, it displays a message in Ajax response visible only in dev tools:

![image](https://user-images.githubusercontent.com/2185289/57152248-92d1d000-6ddb-11e9-8c1b-96273fd02146.png)

And changes are not visible on Preview page, of course.
I have tried to change themes but problem persists.

**Comment by earnjam** [comment]
@arnisjuraga have you tested this with the latest version of the Gutenberg plugin, or with the WordPress 5.2 release candidate?

The fix for this has been merged, but will not be released in core until 5.2 comes out next week.

**Comment by arnisjuraga** [comment]
No, haven't tested. Let's leave it then. Will wait for latest upgrade and will check then.
